### PR TITLE
8274617: constructor and parameter annotations are not copied to the anonymous class constructor

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Please review this PR which is about propagating constructor and parameter annotations to the anonymous class constructor. Currently we are propagating them to bridge methods which are synthetic. Propagating them to the anonymous class constructor seems sensible given that they are mandated by the spec. Please review also the related CSR,

TIA

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed
- [ ] Change requires a CSR request to be approved

### Integration blocker
&nbsp;⚠️ Title mismatch between PR and JBS for issue [JDK-8274617](https://bugs.openjdk.java.net/browse/JDK-8274617)

### Issues
 * [JDK-8274617](https://bugs.openjdk.java.net/browse/JDK-8274617): annotations in the super class constructor are not propagated to the anonymous class constructor ⚠️ Title mismatch between PR and JBS. ⚠️ Issue is not open.
 * [JDK-8277679](https://bugs.openjdk.java.net/browse/JDK-8277679): annotations in the super class constructor are not propagated to the anonymous class constructor (**CSR**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6548/head:pull/6548` \
`$ git checkout pull/6548`

Update a local copy of the PR: \
`$ git checkout pull/6548` \
`$ git pull https://git.openjdk.java.net/jdk pull/6548/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6548`

View PR using the GUI difftool: \
`$ git pr show -t 6548`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6548.diff">https://git.openjdk.java.net/jdk/pull/6548.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java [hunk 1]
```diff
@@ -1337,9 +1337,17 @@ public Type constructorType() {
 
         @Override
         public MethodSymbol constructorSymbol() {
+            // we should do this only once
+            boolean copyConstAnnos = constructorSymbol == null;
```
Comment (lahodaj) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java:1341]: Question: can we avoid repeating this method completely if `constructorSymbol != null`? Seems a bit weird to not copy the annotation again (which I understand), but redo all the rest again.

File: test/langtools/tools/javac/annos_not_copied_anonymous_ctr/AnnosNotCopiedToAnonymousCtrTest.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface VisibleCtrAnnotation {}
+
+    @Target(value = {ElementType.CONSTRUCTOR})
+    @interface InvisibleCtrAnnotation {}
+
+    public class Test {
+        @VisibleCtrAnnotation
+        @InvisibleCtrAnnotation
+        public Test(String firstParam, @VisibleParamAnnotation String secondParam, @InvisibleParamAnnotation String thirdParam) {}
```
Comment (lahodaj) [inline test/langtools/tools/javac/annos_not_copied_anonymous_ctr/AnnosNotCopiedToAnonymousCtrTest.java:116]: How about type annotations of the parameter types? Should these be copied? (And tested?)

File: test/langtools/tools/jdeps/listdeps/ListModuleDeps.java [hunk 3]
```diff
@@ -95,6 +95,7 @@ public void compileAll() throws Exception {
                                 "java.base/jdk.internal.javac",
                                 "java.base/jdk.internal.jmod",
                                 "java.base/jdk.internal.misc",
+                                "java.base/jdk.internal.vm.annotation",
```
Comment (lahodaj) [inline test/langtools/tools/jdeps/listdeps/ListModuleDeps.java:98]: This looks a bit suspicious to me - why the new entry? Do we know why it was necessary? (I would expect that any change here would be caused by a change in dependencies among modules, but I don't seem to see that change.)

File: test/langtools/tools/javac/annos_not_copied_anonymous_ctr/AnnosNotCopiedToAnonymousCtrTest.java (hunk 2 above)
Comment (vicente-romero-oracle) [reply test/langtools/tools/javac/annos_not_copied_anonymous_ctr/AnnosNotCopiedToAnonymousCtrTest.java:116]: yep I should add a test for type annotations, good point

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java (hunk 1 above)
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java:1341]: yep good point

File: test/langtools/tools/javac/annos_not_copied_anonymous_ctr/AnnosNotCopiedToAnonymousCtrTest.java (hunk 2 above)
Comment (vicente-romero-oracle) [reply test/langtools/tools/javac/annos_not_copied_anonymous_ctr/AnnosNotCopiedToAnonymousCtrTest.java]: I have added support for type annotations, thanks for the comments

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java [hunk 4]
```diff
@@ -1337,25 +1340,76 @@ public Type constructorType() {
...
+                ListBuffer<JCVariableDecl> params = new ListBuffer<>();
+                if (csym.params != null) {
+                    for (VarSymbol param : csym.params) {
+                        /* the compiler will issue an error if a type annotation is applied to, for example, `java.lang.String`
+                         * but it won't if the type annotation is applied to `String`, at this point all types will be represented
+                         * in its fully unfolded form so we need to go back to the short form if type annotations are present for a
+                         * given parameter
+                         */
+                        if (!param.getRawTypeAttributes().isEmpty()) {
+                            params.append(make.VarDef(param, make.Ident(param.type.tsym), null));
```
Comment (lahodaj) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java:1390]: I guess I would only use this branch, whether or not there is a type annotation. (Would be much better tested, among other things.)

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java [hunk 5]
```diff
@@ -1337,25 +1340,76 @@ public Type constructorType() {
...
+                            params.append(make.VarDef(param,null));
+                        }
+                    }
+                }
+                return params.toList();
+            }
+
+            private List<JCTypeParameter> typeParams(List<Type> typarams) {
+                ListBuffer<JCTypeParameter> tparams = new ListBuffer<>();
+                final AtomicInteger paramIndex = new AtomicInteger(0);
```
Comment (lahodaj) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java:1401]: The AtomicInteger is a big weird (it is the first time it is used in javac,  believe). I wonder if using some alternative (like an array) would not fit better.

**Comment by liach** [comment]
Can these annotations be detected properly by reflection at runtime (when, say, compiled with `-parameters` flag to avoid a few other existing issues)?

**Comment by vicente-romero-oracle** [comment]
> Can these annotations be detected properly by reflection at runtime (when, say, compiled with `-parameters` flag to avoid a few other existing issues)?

sorry I'm not sure I'm following you, the `-parameters` option deals with the parameter names it has nothing to do with annotations

**Comment by liach** [comment]
Hmm, since both annotation and method signature attributes' recorded signature/annotations don't exactly correspond to the parameters in the descriptor, I thought they probably would share the same mechanism of determining the exact parameter mapping. Reflection uses `MethodParameters` attribute's modifiers to determine how signatures apply to parameters, while it handles annotation in a separate way.

This indeed doesn't matter as the extra local variables passed are in the end of parameter list, which doesn't impact the current way annotations are written to class files and read at runtime. Sorry for digressing here.

**Pull request closed without merging**